File: java/src/main/java/io/epoch/Engine.java
package io.epoch;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public final class Engine {
    private static final long FNV_OFFSET_BASIS = 0xcbf29ce484222325L;
    private static final long FNV_PRIME = 0x100000001b3L;

    private Engine() {
    }

    public static final class Message {
        public final long epoch;
        public final long channelId;
        public final long sourceId;
        public final long sourceSeq;
        public final long schemaId;
        public final long payload;

        public Message(long epoch, long channelId, long sourceId, long sourceSeq, long schemaId, long payload) {
            this.epoch = epoch;
            this.channelId = channelId;
            this.sourceId = sourceId;
            this.sourceSeq = sourceSeq;
            this.schemaId = schemaId;
            this.payload = payload;
        }
    }

    public static final class EpochResult {
        public final long epoch;
        public final long state;
        public final String hash;

        public EpochResult(long epoch, long state, String hash) {
            this.epoch = epoch;
            this.state = state;
            this.hash = hash;
        }
    }

    public static String fnv1a64Hex(String input) {
        long hash = FNV_OFFSET_BASIS;
        byte[] bytes = input.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        for (byte b : bytes) {
            hash ^= (b & 0xffL);
            hash *= FNV_PRIME;
        }
        return toHex(hash);
    }

    public static List<EpochResult> processMessages(List<Message> messages) {
        messages.sort(Comparator
            .comparingLong((Message m) -> m.epoch)
            .thenComparingLong(m -> m.channelId)
            .thenComparingLong(m -> m.sourceId)
            .thenComparingLong(m -> m.sourceSeq));

        List<EpochResult> results = new ArrayList<>();
        long state = 0;
        long currentEpoch = 0;
        boolean hasEpoch = false;

        for (Message message : messages) {
            if (!hasEpoch) {
                currentEpoch = message.epoch;
                hasEpoch = true;
            }
            if (message.epoch != currentEpoch) {
                results.add(new EpochResult(currentEpoch, state, fnv1a64Hex("state:" + state)));
                currentEpoch = message.epoch;
            }
            state += message.payload;
        }

        if (hasEpoch) {
            results.add(new EpochResult(currentEpoch, state, fnv1a64Hex("state:" + state)));
        }

        return results;
    }

    private static String toHex(long value) {
        String hex = Long.toUnsignedString(value, 16);
        if (hex.length() >= 16) {
            return hex;
        }
        StringBuilder sb = new StringBuilder(16);
        for (int i = hex.length(); i < 16; i++) {
            sb.append('0');
        }
        sb.append(hex);
        return sb.toString();
    }
}

File: python/epoch/engine.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, List

FNV_OFFSET_BASIS = 0xCBF29CE484222325
FNV_PRIME = 0x100000001B3
FNV_MASK = 0xFFFFFFFFFFFFFFFF


@dataclass(frozen=True)
class Message:
    epoch: int
    channel_id: int
    source_id: int
    source_seq: int
    schema_id: int
    qos: int
    payload: int


@dataclass(frozen=True)
class EpochResult:
    epoch: int
    state: int
    hash: str


def fnv1a64_hex(value: str) -> str:
    h = FNV_OFFSET_BASIS
    for b in value.encode("utf-8"):
        h ^= b
        h = (h * FNV_PRIME) & FNV_MASK
    return f"{h:016x}"


def process_messages(messages: Iterable[Message]) -> List[EpochResult]:
    ordered = sorted(
        messages,
        key=lambda m: (m.epoch, m.channel_id, -m.qos, m.source_id, m.source_seq),
    )
    results: List[EpochResult] = []
    state = 0
    current_epoch = None

    for message in ordered:
        if current_epoch is None:
            current_epoch = message.epoch
        if message.epoch != current_epoch:
            results.append(
                EpochResult(
                    epoch=current_epoch,
                    state=state,
                    hash=fnv1a64_hex(f"state:{state}"),
                )
            )
            current_epoch = message.epoch
        state += message.payload

    if current_epoch is not None:
        results.append(
            EpochResult(
                epoch=current_epoch,
                state=state,
                hash=fnv1a64_hex(f"state:{state}"),
            )
        )

    return results
